File: day3/p1.py
import re

def extract_valid_mul_instructions(memory):
    # Regex to match only valid mul() instructions with exactly two comma-separated integers
    mul_pattern = r'mul\((\d+),(\d+)\)'
    
    # Find all valid mul() instructions
    valid_muls = re.findall(mul_pattern, memory)
    
    # Convert to integers and multiply
    results = [int(x) * int(y) for x, y in valid_muls]
    
    return results

def solve_corrupted_memory(memory):
    # Extract and multiply results
    mul_results = extract_valid_mul_instructions(memory)
    
    # Sum all multiplication results
    total_sum = sum(mul_results)
    
    return total_sum, mul_results

# Read the input file
with open('input_puzzle.txt', 'r') as file:
    corrupted_memory = file.read()

# Solve the puzzle
total, individual_results = solve_corrupted_memory(corrupted_memory)

print(f"Total sum of multiplications: {total}")
print(f"Number of valid mul() instructions: {len(individual_results)}")

File: day3/p1.js
const fs = require("fs");

// Function to extract and evaluate valid mul() instructions
function extractValidMulInstructions(memory) {
    const mulPattern = /mul\((\d+),(\d+)\)/g; // Regex to match mul() instructions
    let match;
    const results = [];

    // Iterate through all matches
    while ((match = mulPattern.exec(memory)) !== null) {
        const x = parseInt(match[1], 10);
        const y = parseInt(match[2], 10);
        results.push(x * y);
    }

    return results;
}

// Function to solve the corrupted memory
function solveCorruptedMemory(memory) {
    // Extract and multiply results
    const mulResults = extractValidMulInstructions(memory);

    // Calculate the total sum
    const totalSum = mulResults.reduce((sum, val) => sum + val, 0);

    return { totalSum, mulResults };
}

// Read the input file
const filePath = "input_puzzle.txt";
try {
    const corruptedMemory = fs.readFileSync(filePath, "utf8");

    // Solve the puzzle
    const { totalSum, mulResults } = solveCorruptedMemory(corruptedMemory);

    console.log(`Total sum of multiplications: ${totalSum}`);
    console.log(`Number of valid mul() instructions: ${mulResults.length}`);
} catch (err) {
    console.error(`Error reading file: ${err.message}`);
}
